internal/dataplane/parser/translators/httproute_atc.go

```diff
@@ -0,0 +1,216 @@
+package translators
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/kong/go-kong/kong"
+	"github.com/samber/lo"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/annotations"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/kongstate"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser/atc"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+// GenerateKongExpressionRoutesFromHTTPRouteMatches generates Kong routes from HTTPRouteRule
+// pointing to a specific backend.
+func GenerateKongExpressionRoutesFromHTTPRouteMatches(
+	routeName string,
+	matches []gatewayv1beta1.HTTPRouteMatch,
+	filters []gatewayv1beta1.HTTPRouteFilter,
+	ingressObjectInfo util.K8sObjectInfo,
+	hostnames []string,
+	tags []*string,
+) ([]kongstate.Route, error) {
+	// initialize the route with route name, preserve_host, and tags.
+	r := kongstate.Route{
+		Ingress: ingressObjectInfo,
+		Route: kong.Route{
+			Name:         kong.String(routeName),
+			PreserveHost: kong.Bool(true),
+			Tags:         tags,
+		},
+		ExpressionRoutes: true,
+	}
+
+	if len(matches) == 0 {
+		if len(hostnames) == 0 {
+			return []kongstate.Route{}, ErrRouteValidationNoMatchRulesOrHostnamesSpecified
+		}
+
+		hostMatcher := hostMatcherFromHosts(hostnames)
+		atc.ApplyExpression(&r.Route, hostMatcher, 1)
+		return []kongstate.Route{r}, nil
+	}
+
+	_, hasRedirectFilter := lo.Find(filters, func(filter gatewayv1beta1.HTTPRouteFilter) bool {
+		return filter.Type == gatewayv1beta1.HTTPRouteFilterRequestRedirect
+	})
+	// if the rule has request redirect filter(s), we need to generate a route for each match to
+	// attach plugins for the filter.
+	if hasRedirectFilter {
+		routes := make([]kongstate.Route, 0, len(matches))
+		for _, match := range matches {
+			matchRoute := kongstate.Route{
+				Ingress: ingressObjectInfo,
+				Route: kong.Route{
+					Name:         kong.String(routeName),
+					PreserveHost: kong.Bool(true),
+					Tags:         tags,
+				},
+				ExpressionRoutes: true,
+			}
+			// generate matcher for this HTTPRoute Match.
+			matcher := atc.And(generateMatcherFromHTTPRouteMatch(match))
+
+			// add matcher from parent httproute (hostnames, protocols, SNIs) to be ANDed with the matcher from match.
+			matchersFromParent := matchersFromParentHTTPRoute(hostnames, ingressObjectInfo.Annotations)
+			for _, m := range matchersFromParent {
+				matcher.And(m)
+			}
+			atc.ApplyExpression(&matchRoute.Route, matcher, 1)
+
+			// we need to extract the path to configure redirect path of the plugins for request redirect filter.
+			path := ""
+			if match.Path != nil && match.Path.Value != nil {
+				path = *match.Path.Value
+			}
+			plugins := GeneratePluginsFromHTTPRouteFilters(filters, path, tags)
+			matchRoute.Plugins = plugins
+
+			routes = append(routes, matchRoute)
+		}
+		return routes, nil
```

Comment:
How about extracting a separate function from this code block to limit the indentation and complexity?
```suggestion
		return generateKongExpressionRoutesWithRedirectFilter(
			routeName,
			matches,
			filters,
			ingressObjectInfo,
			hostnames,
			tags,
		)
```